File: Python/Sorting/merge_sort.py
# Function that performs the sorting operation
def merge(arr, l, mid, r):
    i, j = l, mid+1
    b = []
    # Compares the elements of the two arrays and sorts them
    while i<=mid and j<=r:
        if arr[i]>arr[j]:
            b.append(arr[j])
            j+=1
        else:
            b.append(arr[i])
            i+=1
    # If the 1st array is exhausted, then copies the remaining elements of the 2nd array to the 1st array
    while j<=r:
        b.append(arr[j])
        j+=1
    
    # If the 2nd array is exhausted, then copies the remaining elements of the 1st array to the 2nd array
    while i<=mid:
        b.append(arr[i])
        i+=1

    # Copies all the elements of the array 'b' to array 'arr'
    for i in range(l, r+1):
        arr[i] = b[i-l]

    return arr

# Function that divides the array into half at each step and sorts them
def mergeSort(arr, l, r):
    if l<r:
        mid = int(l + (r-l)/2)
        # Sorts the left array
        arr = mergeSort(arr, l, mid)
        # Sorts the right array
        arr = mergeSort(arr, mid+1, r)
        # Merges the left and the right array
        arr = merge(arr, l, mid, r)
    return arr

# Total number of elements in vector
n = int(input())
arr = []
for i in range(n):
    x = int(input())
    arr.append(x)

arr = mergeSort(arr, 0, n-1)
print("After sorting:")
for i in arr:
    print(i, end=" ")

File: C++/Sorting/merge_sort.cpp
#include <bits/stdc++.h>

using namespace std;

// Function that performs the sorting operation
void merge(vector<int>& v, int l, int mid, int r){
    int i = l, j = mid+1;
    vector<int> b;
    // Compares the elements of the two arrays and sorts them
    while (i<=mid && j<=r)
    {
        if (v[i]>v[j]){
            b.push_back(v[j]);
            j++;
        }
        else{
            b.push_back(v[i]);
            i++;
        }
    }
    // If the 1st array is exhausted, then copies the remaining elements of the 2nd array to the 1st array
    while (j<=r){
        b.push_back(v[j]);
        j++;
    }
    // If the 2nd array is exhausted, then copies the remaining elements of the 1st array to the 2nd array
    while (i<=mid){
        b.push_back(v[i]);
        i++;
    }
    // Copies all the elements of the array 'b' to array 'arr'
    for (int i = l; i <= r; i++){
        v[i] = b[i-l];
    }
    return ;
}

// Function that divides the array into half at each step and sorts them
void mergeSort(vector<int>& v, int l, int r){
    if (l<r)
    {
        int mid = l + (r-l)/2;
        // Sorts the left array
        mergeSort(v, l, mid);
        // Sorts the right array
        mergeSort(v, mid+1, r);
        // Merges the left and the right array
        merge(v, l, mid, r);
    }
}

int main()
{
    // Total number of elements in vector
    int n;
    cin >> n;
    vector<int> v(n);
    for(int i=0; i<n; i++){
        cin >> v[i];
    }
    mergeSort(v, 0, n-1);
    cout << "After sorting:\n";
    for(auto i: v){
        cout << i << " ";
    }
    return 0;
}
